racket/src/rktio/rktio_fs.c

```diff
@@ -701,6 +702,94 @@ rktio_ok_t rktio_set_current_directory(rktio_t *rktio, const char *path)
   return !err;
 }
 
+rktio_stat_t *rktio_file_or_directory_stat(
+  rktio_t *rktio, rktio_const_string_t path, rktio_bool_t follow_links)
+{
+  int stat_result;
+  struct rktio_stat_t *rktio_stat_buf;
+#ifdef RKTIO_SYSTEM_UNIX
+  struct stat stat_buf;
+
+  do {
+    if (follow_links) {
+      stat_result = stat(path, &stat_buf);
+    } else {
+      stat_result = lstat(path, &stat_buf);
+    }
+  } while ((stat_result == -1) && (errno == EINTR));
+
+  if (stat_result) {
+    get_posix_error();
+    return NULL;
+  } else {
+    rktio_stat_buf = (struct rktio_stat_t *) malloc(sizeof(struct rktio_stat_t));
+    rktio_stat_buf->device_id = stat_buf.st_dev;
+    rktio_stat_buf->inode = stat_buf.st_ino;
+    rktio_stat_buf->mode = stat_buf.st_mode;
+    rktio_stat_buf->hardlink_count = stat_buf.st_nlink;
+    rktio_stat_buf->user_id = stat_buf.st_uid;
+    rktio_stat_buf->group_id = stat_buf.st_gid;
+    rktio_stat_buf->device_id_for_special_file = stat_buf.st_rdev;
+    rktio_stat_buf->size = stat_buf.st_size;
+    rktio_stat_buf->block_size = stat_buf.st_blksize;
+    rktio_stat_buf->block_count = stat_buf.st_blocks;
+    /* The `tv_nsec` fields are only the fractional part of the seconds.
+       (The value is always lower than 1_000_000_000.) */
+    rktio_stat_buf->access_time_seconds = stat_buf.st_atim.tv_sec;
+    rktio_stat_buf->access_time_nanoseconds = stat_buf.st_atim.tv_nsec;
+    rktio_stat_buf->modify_time_seconds = stat_buf.st_mtim.tv_sec;
+    rktio_stat_buf->modify_time_nanoseconds = stat_buf.st_mtim.tv_nsec;
+    rktio_stat_buf->ctime_seconds = stat_buf.st_ctim.tv_sec;
+    rktio_stat_buf->ctime_nanoseconds = stat_buf.st_ctim.tv_nsec;
+    rktio_stat_buf->ctime_is_change_time = 1;
+    return rktio_stat_buf;
+  }
+#endif
+#ifdef RKTIO_SYSTEM_WINDOWS
+  struct _wstat64 stat_buf;
+  const WIDE_PATH_t *wp;
+  wp = MSC_WIDE_PATH_temp(path);
+  if (!wp) {
+    return NULL;
+  }
+
+  do {
+    /* No stat/lstat distinction under Windows */
+    stat_result = _wstat64(wp, &stat_buf);
+  } while ((stat_result == -1) && (errno == EINTR));
+
+  if (stat_result) {
+    /* TODO: Error handling / integration with Racket side? */
+    return NULL;
+  }
+
+  rktio_stat_buf = (struct rktio_stat_t *) malloc(sizeof(struct rktio_stat_t));
```

Comment:
Since this block of code is (almost?) the same as the Unix one, would it make sense to share more?

Reply:
I wrote something about this in the commit message for [5466b9f9]. You may have missed it because it was further down:
```
    The sections for Posix and Windows that copy values from `stat_buf` to
    `rktio_stat_buf` are similar, but not the same:
    
    - Some comments refer only to Posix, some only to Windows.
    - The struct fields for timestamps in `stat_buf` are different for
      Posix and Windows.
    
    Though it would be possible to put common code lines in a section
    outside the `ifdef`s, it could well be that this refactoring makes the
    code _more_ difficult to follow. Also keep in mind that we just copy
    data from stat structs that are unlikely to change, due to backward
    compatibility considerations. So we might not benefit from being able to
    change common code in one place. Also it might be that the "common" code
    needs to be changed in different ways for Posix and Windows.
    
    Long story short: Probably it's better to keep the Posix and Windows
    implementations separate, but I leave it to you.
```